### ISSUE TITLE
Wrong collection type in Set assertion

### ISSUE THREAD

**Issue by fbiville** [assertj/assertj-core#2132]
#### Summary

The assertion error displays the wrong collection type, as noticed by @Patouche.

#### Example

```java
import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

class AssertJTests {

    @Test
    void iterable_bad_name() {
        final Set<String> actual = Stream.of("toto", "tata").collect(Collectors.toSet());
        Assertions.assertThat(actual).contains("tutu");
    }
}
```

The error is as follows:
```java
java.lang.AssertionError: 
Expecting ArrayList:
 <["toto", "tata"]>
to contain:
 <["tutu"]>
but could not find the following element(s):
 <["tutu"]>
```

`ArrayList` should not be displayed in the error message.

**Comment by scordio** [comment]
The issue happens due to:

https://github.com/assertj/assertj-core/blob/8912e7324fab207e2863891d499cd111eb10f129/src/main/java/org/assertj/core/internal/Iterables.java#L342

introduced in a06f16c500c317fb7f07c45c73c6c6db4e848058 to address non-restartable iterables.

The wrong error message can probably be fixed with some refactoring, but I am wondering if creating a copy makes sense in case of `Collection` instances.

@assertj/core thoughts?

**Comment by Patouche** [comment]
I thinks it make sense as an iterator can be consume only once.

The simplest way to solve this issue will be probably to provide the class to the `org.assertj.core.error.ShouldContain#shouldContain` method.

If you want, I already solve the test I've written before .

**Comment by scordio** [comment]
Thanks @Patouche, what I meant is with `Collection` instances we probably don't need to use an iterator and create a new array list to perform the `contains` assertion, we could simply delegate to `Collection#contains`

**Comment by Patouche** [comment]
Mmmh, not sure we can do that with the `ComparisonStrategy`.

The `Collection#contains` method will use the `equals` methods of the iterable elements. The `Iterables#comparisonStrategy` will not be used anymore if we change that with `contains` method.

**Comment by joel-costigliola** [comment]
@Patouche I agree with you.
The change was made to support iterable that can be traversed only once by an iterator (a use case a bit far fetched but why not).

Passing the actual type to the error message is a good solution IMO.

**Comment by scordio** [comment]
> Passing the actual type to the error message is a good solution IMO. WDYT @scordio ?

Looks good to me, we might still improve the part of the copy for performance reasons. I'll have a look at it.

**Comment by Patouche** [comment]
> The change was made to support iterable that can be traversed only once by an iterator (a use case a bit far fetched but why not).

Yes, I understand the use case but I wonder how this can work when chaining some assertions. For example :

```java
class SampleTests {
    @Test
    void sample() {
        Iterable<String> iterable = createSinglyIterable("val-1", "val-2")
        assertThat(iterable)
            .contains("val-1")
            .contains("val-2");
    }
}
```

Maybe I'm wrong but I think this should fail on the `contains("val-2")` line as the iterator was already consumed on the previous line.

> Looks good to me, we might still improve the part of the copy for performance reasons. I'll raise a separate issue to look into it.

Yes, this is a little bit out of the scope of this PR but maybe instead of `newArrayList` we can use `IterableUtil.toCollection` in the `Iterables#contains` which will prevent the copy you're talking when the iterable is already a `Collection`. In all others cases, an `ArrayList` may be recreated.

**Comment by fbiville** [comment]
```java
class SampleTests {
    @Test
    void sample() {
        Iterable<String> iterable = createSinglyIterable("val-1", "val-2")
        assertThat(iterable)
            .contains("val-1")
            .contains("val-2");
    }
}
```
There could be ways to have this assertion pass (since you don't have to iterate until the end to find the first value.
But I think the following should fail for sure:
```java
class SampleTests {
    @Test
    void sample() {
        Iterable<String> iterable = createSinglyIterable("val-1", "val-2")
        assertThat(iterable)
            .contains("val-2")
            .contains("val-1");
    }
}
```
#justmy2cents

**Comment by scordio** [comment]
@fbiville would you mind elaborating on why the second example should fail?

Although I understand the single iteration constraint, I'm concerned that the order of the assertions would lead to different results which might be nonobvious from user perspective.

Specifically, following your example:

```java
assertThat(createSinglyIterable("val-1", "val-2")).contains("val-1"); // succeeds

assertThat(createSinglyIterable("val-1", "val-2")).contains("val-2"); // succeeds

assertThat(createSinglyIterable("val-1", "val-2"))
  .contains("val-1"); // succeeds
  .contains("val-2"); // succeeds or fails?

assertThat(createSinglyIterable("val-1", "val-2"))
  .contains("val-2"); // succeeds
  .contains("val-1"); // fails

assertThat(createSinglyIterable("val-1", "val-2")).contains("val-1", "val-2"); // succeeds

assertThat(createSinglyIterable("val-1", "val-2")).contains("val-2", "val-1"); // succeeds or fails?
```

It might be we should just copy all the elements in a collection during `assertThat` execution in case the input is an `Iterable`, so no assumption is taken about the underlying implementation. I tend to think that the side effect (if there is a need for it) should be consistent and not depending on how the assertions are written/executed in the test. 

As mentioned before, we are talking about a niche use case so I'm actually wondering if we should take any action to address it, considering that there is no issue raised about the current behavior.
